### DISCUSSION TITLE
[back] Dm 생성과 삭제 관련 미설계 부분 토론 요청

### DISCUSSION THREAD

**Discussion by kyu-baek** [3DPong/transcendence#88] [Q&A]
#### Dm 은 두가지 테이블과 연관이 있습니다.

- 첫 번째는 Chat_Channel (채널의 대한 정보) 

<img width="500" alt="image" src="https://user-images.githubusercontent.com/69143394/229837927-d368abf8-ae1f-4182-a153-4fdcaa0b3cf3.png">

- 두 번째는 Dm_Channel(채널에 포함된 **유저에** 대한 정보) 입니다. 

<img width="500" alt="image" src="https://user-images.githubusercontent.com/69143394/229836572-40cd0633-0ac1-45c2-bc94-80b0654cb380.png">

Dm 채널을 create 하는 경우는 일방 그룹 채팅방을 생성할 때와 비슷한 흐름입니다.

하지만 삭제의 경우는 조금 복잡해집니다.

그룹 채팅방은 하나의 채팅방에 포함된 유저의 인원 수 만큼 channel_user 테이블이 추가되는 반면, 
dm 은 오직 두 사람 만을 위한 채팅 방이기 때문에 
하나의 채팅방에 한 개의 Dm_Channel 테이블이(first_user, second_user, channel_id ...) 추가됩니다. 

누군가 그룹채팅 방을 나가면, 그 채팅방 안에 남아있는 유저가 있는 한 Chat_Channel 은 soft-delete 되지 않습니다.
dm에서 둘 중 한사람이 나가기를 하면, Chat_Channel 과 Dm_Channel 데이터를 지우기도 애매하고 그냥 두기도 애매한 것 같습니다. 
Dm_Channel 컬럼은 지울 수 없는 게 둘 중 한명의 유저 정보가 있기 때문이고, 
아예 지우는 행위를 안하면, 해당유저는 나가기한 dm 채팅방이 리스트에서 재검색 되는 문제가 있습니다. 

first_user_id, second_user_id 가 서로 pk 로 묶여 있기 때문에 하나면 null 시키는 것도 불가능 할 것 같기 때문에,

#### 제 해결 방법은 

1. 한명만 나가도 둘다 삭제해버리기 -> 다른 유저는 억울하게 데이터 지워지는 단점이 있지만, dm 은 추가기능일 뿐입니당

2. first_user_status, second_user_status 를 추가하고 true, false 로 구별하고,
만약 나가기한 유저가 false 상태지만, 나중에 다시 같은 유저에게 dm 을 요청하면, 
이전 내용까지 다시 채팅방이 불려올 수 있게끔 설정하는 겁니다.

다른 아이디어가 있다면 카톡이나 댓글로 써주세용.

**Comment by parktest0325** [comment]
이렇게 구현했을때 발생하는 문제점이 있을까요? 

일단 DM도 일반 채널과 비슷하게 관리

### DM 보내기 버튼
1. DM 채널이 이미 있는경우?
  1-1. 안나간경우 => 그냥 그거 불러옴
  1-2. 나가기를 한 경우 => dm_channel에서 채널을 찾을 수 있음 => 해당 채널의 channel_user에 자기자신 추가
2. 없는경우?
  그냥 생성 후 channel_user, dm_channel 에 데이터 추가 


### DM 채널에서 메시지 보내기
1. 상대방이 안나간경우
  그냥 일반 채널처럼 메시지 보내기
2. 상대방이 나간경우
  상대방을 channel_user에 추가 후 메시지보내기
상대방이 나갔는지 안나갔는지 검사가 필요하기 때문에 메시지 보낼때마다 계속해서 emit전에 channel_user 확인 필요
또는 메시지 타입을 DM용 타입을 만들어서 메시지 보내기 전 검증과정 거침


### 나가기
1. 한명 나간경우 channel_user만 지우고 그냥 그대로 둠 어차피 메시지 보낼때 초대됨
2. 두번째로 나간경우 channel_user만 지우고 channel, dm_channel은 그대로 둠

**Comment by kyu-baek** [comment]
상대방을 channel_user에 추가 후 메시지보내기 -> 메세지 하나를 보낼 때마다 상대방에 나갔는지 안나갔는지 검사하는 건 효율적이지 않은 것 같습니다.. 그냥 혼자 말하게 하는 걸로 두는 것도 나쁘지 않은 것 같습니다. 어차피 옆에 유저 목록이 있으니 혼자 있는지 아닌지 알 수 있다고 생각합니다.

**Comment by JuneParkCode** [comment]
첫 메시지를 보내는 경우와 비슷한 것 같은데요.

chat user 에서는 지우되, dm user에서는 남겨두고,

상대방이 새로운 메시지를 보내면 chat user쪽으로 다시 불러들여야 할 것 같습니다.

말씀대로 상대방이 채팅에 존재하는지 아닌지 검증이 조금 문제일 것 같긴한데... 이건 조금만 더 생각해보겠습니다..

**Reply by kyu-baek** [reply]
메세지를 보낼때마다, 상대방이 있는지 없는지 확인하는 게 비효율적인 것 같아서요. 나가는 걸 아예 막는 방식을 생각하고 있습니다.

**Comment by JuneParkCode** [comment]
규림님이 원하시는 작동은 이런걸까요?

1. DM 방에서 나간다.
2. 나간 사용자의 뷰에서는 나오지 않는다. 하지만 상대방은 그대로.
3. 새로운 메시지가 오면 다시 나타난다 (?)
OR 새로운 메시지를 상대방이 보낼 수 없다?

3번에서 어떻게 작동하냐에 따라서 어떻게 될 지 다를 것 같습니다.

**Comment by kyu-baek** [comment]
dm 은 나가기 없이 가는 방향은 어떤가요?
대신 슬랙처럼 최신 5개정도만 보여주고, 나중에 친구의 디엠을 다시 걸면 기존의 메세지도 다시 불러 올 수 있는 거죠.

**Reply by JuneParkCode** [reply]
이게 젤 현명한 방법 같네요

**Reply by parktest0325** [reply]
이 방식이면 이제 dm 을 상단에 표시되도록 정렬해도 되겠네요

**Reply by kyu-baek** [reply]
> 이 방식이면 이제 dm 을 상단에 표시되도록 정렬해도 되겠네요

그럼 list api 를 dm 으로 따로 요청해서 받아오는 게 바로 정렬하기 편하실 수도 있으니 필요하면 말씀해주세요.
지금 현재는 dm 타입 5개가 맨 뒤 5개로 설정되어 있습니다. 그럼 차라리 앞에 5개로 넣어주면 받아온 list 그대로 뿌려주기만 해도 dm이 먼저 정렬되는 방식이 될 것 같습니다. 편하신 방식 선택해서 알려주세요.
